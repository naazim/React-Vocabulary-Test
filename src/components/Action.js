import React from 'react';
import Progress from './Progress';

const Action = (props) => (
  <div>
    <Progress percentage={props.percentage} />
    <button
      className="big-button"
      onClick={props.handleTest}
      disabled={!props.hasOptions}
    >
      Take Test
      </button>
  </div>
);

export default Action;
