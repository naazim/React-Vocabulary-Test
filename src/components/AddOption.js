import React from 'react';

export default class AddOption extends React.Component {
  state = {
    error: undefined
  };
  handleAddOption = (e) => {
    e.preventDefault();
    const option = e.target.elements.option.value.trim();
    const option2 = e.target.elements.option2.value.trim();
    const error = this.props.handleAddOption(option, option2);

    this.setState(() => ({ error }));

    if (!error) {
      e.target.elements.option.value = '';
      e.target.elements.option2.value = '';
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p className="add-option-error">{this.state.error}</p>}
        <form className="add-option" onSubmit={this.handleAddOption}>
          <input className="add-option__input" type="text" name="option" placeholder="Enter native word" />
          <input className="add-option__input" type="text" name="option2" placeholder="Enter foreign word" />
          <button className="button">Add Option</button>
        </form>
      </div>
    );
  }
}
