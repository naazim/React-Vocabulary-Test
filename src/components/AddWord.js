import React from 'react';

export default class AddWord extends React.Component {

  state = {
    error: undefined
  };

  handleAddWord = (e) => {
    e.preventDefault();
    const native = e.target.elements.native.value.trim();
    const foreign = e.target.elements.foreign.value.trim();
    const error = this.props.handleAddWord(native, foreign);

    this.setState(() => ({ error }));

    if (!error) {
      e.target.elements.native.value = '';
      e.target.elements.foreign.value = '';
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p className="add-word-error">{this.state.error}</p>}
        <form className="add-word" onSubmit={this.handleAddWord}>
          <input className="add-word__input" type="text" name="native" placeholder="Enter native word" autoFocus />
          <input className="add-word__input" type="text" name="foreign" placeholder="Enter foreign word" />
          <button className="button button--plus"><span className="add-word__mobile">Add word </span>+</button>
        </form>
      </div>
    );
  }
}
