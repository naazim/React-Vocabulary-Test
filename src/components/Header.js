import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = (props) => (
  <div className="header">
    <div className="container">
      <h1 className="header__title">{props.title}</h1>
      <NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink>
      <NavLink to="/quiz" activeClassName="is-active" exact={true}>Take Test</NavLink>
    </div>
  </div>
);

Header.defaultProps = {
  title: 'Vocabulary Test'
};

export default Header;
