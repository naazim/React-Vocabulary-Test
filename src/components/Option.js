import React from 'react';

const Option = (props) => (
  <div className="option">
      <span className="option__text">{props.nativeWord}</span>
      <span className="option__text">{props.foreignWord}</span>
    <button
      className="button button--icon"
      onClick={(e) => {
        props.handleDeleteWord(props.count - 1);
      }}
    >
      âœ–
      </button>
  </div>
);

export default Option;
