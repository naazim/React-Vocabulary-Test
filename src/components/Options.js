import React from 'react';
import Option from './Option';

const Options = (props) => (
  <div>
    <div className="widget__header">
      <h3 className="widget__title">Your word list</h3>
      <button
        className="button--link"
        onClick={props.handleDeleteAll}
      >
        Remove All
    </button>
    </div>

    {props.options.length === 0 && <p className="widget__message">Please create Vocabulary list to get started!</p>}
    {
      props.options.map((option, index) => {
        return (
          <Option
            key={index}
            nativeWord={option.native}
            foreignWord={option.foreign}
            count={index + 1}
            handleDeleteWord={props.handleDeleteWord}
          />
        )
      })
    }
  </div>
);

export default Options;
