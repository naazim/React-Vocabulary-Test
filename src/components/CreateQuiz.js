import React from 'react';
import Options from './Options';
import AddWord from './AddWord';
import Progress from './Progress';

const CreateQuiz = (props) => {

    /**
     * @description Handles the new word pair added and supplies the error messages
     */
    const handleAddWord = (native, foreign) => {
        if (!native && !foreign) {
            return 'Enter valid words';
        }
        else if (!native) {
            return 'Enter valid native word';
        }
        else if (!foreign) {
            return 'Enter valid foriegn word';
        }

        props.handleAddWord(native, foreign);
    };

    const { wordList } = props;
    const percentage = wordList.length /20 * 100;

    return (
        <div className="wrapper">
            <Progress percentage={percentage} />
            <div className="widget">
                <Options
                    options={wordList}
                    handleDeleteAll={props.handleDeleteAll}
                    handleDeleteWord={props.handleDeleteWord}
                />
                <AddWord
                    handleAddWord={handleAddWord}
                />
            </div>
            <input
                type="button"
                onClick={() => props.handleModule('PerformQuiz')}
                className="button button--big"
                disabled={wordList.length < 20}
                value="Take Test"
            />
        </div>
    );
};

export default CreateQuiz;