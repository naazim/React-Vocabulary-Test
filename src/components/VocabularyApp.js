import React from 'react';
import CreateQuiz from './CreateQuiz';
import Header from './Header';
import Quiz from './Quiz';
import Result from './Result';

export default class VocabularyApp extends React.Component {
  state = {
    wordList: [],
    shuffledList: [],
    module: 'CreateQuiz'  //CreateQuiz || PerformQuiz || QuizResult
  };

  componentWillMount() {
    try {
      const json = localStorage.getItem('wordList');
      const wordList = JSON.parse(json);

      if (wordList) {
        this.setState(() => ({ wordList }));
      }
    } catch (e) {
      // Do nothing
    }
  };

  componentDidUpdate(prevProps, prevState) {
    if (prevState.wordList.length !== this.state.wordList.length) {
      const json = JSON.stringify(this.state.wordList);
      localStorage.setItem('wordList', json);
    };
  };

  /**
   * @description Adds the word pair to the list
   * @param {string} native native word
   * @param {string} foreign translated foreign word
   */
  handleAddWord = (native, foreign) => {
    this.setState((prevState) => ({
      wordList: [...prevState.wordList, { native, foreign }]
    }));
  };

   /**
   * @description Removes the word pair from the list
   * @param {number} wordToRemove index of word pair to be removed from the list
   */
  handleDeleteWord = (wordToRemove) => {
    this.setState((prevState) => ({
      wordList: prevState.wordList.filter((word) => wordToRemove !== prevState.wordList.indexOf(word))
    }));
  };

  /**
   * @description Removes all the word pairs from the list
   */
  handleDeleteAll = () => {
    this.setState(() => ({ wordList: [] }));
  };

  /**
   * @description Switches to the corresponding module / screen
   * @param {string} module name of the module
   */
  handleModule = (module) => {
    this.setState(() => ({
      module
    }));

    if(module === 'PerformQuiz') {  //shuffle for performing quiz only
      const shuffledList = this.shuffleList([...this.state.wordList]);
      this.handleShuffledList(shuffledList);
    }
  };

  /**
   * @description updates the shuffled list
   * @param {array} shuffledList shuffled word list
   */
  handleShuffledList = (shuffledList) => {
    this.setState(() => ({
      shuffledList
    }));
  };

  /**
   * @description Shuffles the array
   * @param {array} list List of word pairs
   * @returns {array} Shuffled list of word pairs
   */
  shuffleList = (list) => {
    for (let i = list.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [list[i], list[j]] = [list[j], list[i]];
    }
    return list;
  };

  render() {
    const { module, wordList, shuffledList } = this.state;

    return (
      <div>
        <Header />
        <div className="container">

          {module === 'CreateQuiz' &&
            <CreateQuiz
              handleAddWord={this.handleAddWord}
              handleDeleteWord={this.handleDeleteWord}
              handleDeleteAll={this.handleDeleteAll}
              handleModule={this.handleModule}
              wordList={wordList}
            />
          }

          {module === 'PerformQuiz' &&
            <Quiz
              handleModule={this.handleModule}
              shuffledList={shuffledList}
              handleShuffledList={this.handleShuffledList}
            />
          }

          {module === 'QuizResult' &&
            <Result
              handleModule={this.handleModule}
              shuffledList={shuffledList}
            />
          }
        </div>
      </div>
    );
  }
}
