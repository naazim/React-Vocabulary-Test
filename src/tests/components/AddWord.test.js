import React from 'react';
import { shallow } from 'enzyme';
import AddWord from '../../components/AddWord';
import {handleAddWord} from '../../components/VocabularyApp';

test('should handle error for invalid form submission', () => {
    const handleAddWord = jest.fn();
    const wrapper = shallow(<AddWord handleAddWord ={handleAddWord} />);

    wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
        target: { elements :{
            native: { value: 'hello' },
            foreign: { value: 'hallo' }
        }}
    });

    expect(handleAddWord).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
});

test('should show error message for invalid form submission', () => {
    const handleAddWord = jest.fn();
    const wrapper = shallow(<AddWord handleAddWord ={handleAddWord} />);
    wrapper.setState({ error: 'Enter valid native word'});
    
    const errorMsg = wrapper.find('.add-word-error').text();
    expect(errorMsg).toEqual('Enter valid native word');
});