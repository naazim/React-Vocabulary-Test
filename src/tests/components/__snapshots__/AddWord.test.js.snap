// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle error for invalid form submission 1`] = `
<div>
  <form
    className="add-word"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="add-word__input"
      name="native"
      placeholder="Enter native word"
      type="text"
    />
    <input
      className="add-word__input"
      name="foreign"
      placeholder="Enter foreign word"
      type="text"
    />
    <button
      className="button button--plus"
    >
      <span
        className="add-word__mobile"
      >
        Add word 
      </span>
      +
    </button>
  </form>
</div>
`;
